src/main/java/org/maxgamer/quickshop/Listeners/BlockListener.java

```diff
@@ -167,19 +169,35 @@ public void onInventoryMove(InventoryMoveItemEvent event) {
             return;
         }
 
-        final Location loc = event.getDestination().getLocation();
-
-        if (loc == null) {
+        if (!plugin.getConfig().getBoolean("shop.update-sign-when-inventory-moving", true)) {
             return;
         }
 
-        final Shop shop = plugin.getShopManager().getShopIncludeAttached(loc);
+        final Inventory inventory = event.getDestination();
+        final Location location = inventory.getLocation();
 
-        if (shop == null) {
+        if (location == null) {
             return;
         }
 
-        shop.setSignText();
+        final Shop shop = plugin.getShopManager().getShop(location);
+        final Shop otherShop = plugin.getShopManager().getShopIncludeAttached(location);
+
+        // Delayed task. Event triggers when item is moved, not when it is received.
+        new BukkitRunnable() {
+            @Override
+            public void run() {
+                if (shop != null) {
+                    shop.setSignText();
+                }
+
+                if (otherShop != null) {
+                    otherShop.setSignText();
+                }
+
+            }
+        }.runTaskLater(plugin, 20);
```

Comment:
This code will create a lot runnable when server have a lot hoppers to shops, it may cause memory leak.
Use queue to replace new Runnable.
@portlek